### PR TITLE
Create library for converting pytorch to caffe2 and examples

### PR DESCRIPTION
**Pull request by houseroad**
Base on this example, we can create notebooks which explain how to compare the results and performance.

### CODE REVIEW

File: onnx_caffe2/pytorch_caffe2.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+import onnx
+import time
+import torch
+import torch.onnx
+
+
+log = logging.getLogger(__name__)
+
+
+def run_caffe2_model(init_net, predict_net, inputs):
```
Comment (bddppq) [inline onnx_caffe2/pytorch_caffe2.py]: Already has one here: https://github.com/onnx/onnx-caffe2/blob/master/onnx_caffe2/helper.py#L56

File: onnx_caffe2/pytorch_caffe2.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    caffe2_results = run_caffe2_model(init_net, predict_net, inputs)
+    if isinstance(pytorch_results, Variable):
+        pytorch_results = [pytorch_results]
+    for p, c in zip(pytorch_results, caffe2_results):
+        np.testing.assert_almost_equal(
+            p.data.cpu().numpy(), c, decimal=expected_decimal)
+
+
+def run_caffe2_benchmark(init_net, predict_net, warmup_iters, main_iters, layer_details):
+    workspace.ResetWorkspace()
```
Comment (bddppq) [inline onnx_caffe2/pytorch_caffe2.py]: Don't directly use `caffe2.python.workspace`, use `onnx_caffe2.workspace.Workspace` instead.

File: onnx_caffe2/pytorch_caffe2.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        te = time.time()
+        total_pytorch_time = te - ts + total_pytorch_time
+    log.info("The PyTorch model execution time per iteration is {} milliseconds."
+                 .format(total_pytorch_time / main_iters * 1000))
+    log.info("Next step is Caffe2.")
+    run_caffe2_benchmark(init_net, predict_net, warmup_iters, main_iters, layer_details)
+    return init_net, predict_net
+
+
+def load_caffe2_net(file):
```
Comment (bddppq) [inline onnx_caffe2/pytorch_caffe2.py]: Maybe move this into helper.py

File: onnx_caffe2/pytorch_caffe2.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    net = caffe2_pb2.NetDef()
+    with open(file, "rb") as f:
+        net.ParseFromString(f.read())
+    return net
+
+
+def save_caffe2_net(net, file, output_txt=False):
```
Comment (bddppq) [inline onnx_caffe2/pytorch_caffe2.py]: helper.py

File: onnx_caffe2/pytorch_caffe2.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    f = io.BytesIO()
+    pytorch_out = torch.onnx._export(model, inputs, f, verbose=verbose, training=training)
+    onnx_model = onnx.ModelProto.FromString(f.getvalue())
+    caffe2_inputs = [var.data.numpy() for var in inputs]
+    init_net, predict_net = Caffe2Backend.onnx_graph_to_caffe2_net(onnx_model.graph)
+    verify_caffe2_model_with_pytorch_results(init_net, predict_net, caffe2_inputs,
+                                             pytorch_out, expected_decimal)
+    log.info("The converted Caffe2 model achieves {}-decimal precision."
+                 .format(expected_decimal))
+    if not compare_performance:
```
Comment (bddppq) [inline onnx_caffe2/pytorch_caffe2.py]: Split this into a separate benchmark/profile function

File: onnx_caffe2/pytorch_caffe2.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        return init_net, predict_net
+
+    log.info("Starting benchmarking PyTorch.")
+    for _i in range(warmup_iters):
```
Comment (bddppq) [inline onnx_caffe2/pytorch_caffe2.py]: May be put these into a `run_pytorch_benchmark` function?

**Comment by houseroad** [comment]
Refactoring the code, will move most of the functions to onnx_caffe2/helper.py. And move the example to the tutorial repo.

File: onnx_caffe2/helper.py [hunk 7]
```diff
@@ -60,17 +68,65 @@ def c2_native_run_net(init_net, predict_net, inputs):
...
+            ws.FeedBlob(key, value, predict_net.device_option)
 
     ws.RunNetOnce(predict_net)
 
     output_names = predict_net.external_output
     output_values = [ws.FetchBlob(name) for name in output_names]
     return ws, namedtupledict('Outputs', output_names)(*output_values)
+
+
+def name_inputs(onnx_model, inputs):
```
Comment (dzhulgakov) [inline onnx_caffe2/helper.py]: you should have the names of blobs in predict_net.external_input. So maybe just pass the list directly to c2_native_run_net?

File: onnx_caffe2/helper.py [hunk 8]
```diff
@@ -60,17 +68,65 @@ def c2_native_run_net(init_net, predict_net, inputs):
...
+        f.write(net.SerializeToString())
+    if output_txt:
+        with open(file + "txt", "w") as f:
+            f.write(str(net))
+
+
+def benchmark_pytorch_model(model, inputs, training=False, warmup_iters=3,
+                            main_iters=10, verbose=False):
+    for _i in range(warmup_iters):
+        ts = time.time()
```
Comment (dzhulgakov) [inline onnx_caffe2/helper.py]: no need for time in warmup

File: onnx_caffe2/helper.py [hunk 9]
```diff
@@ -60,17 +68,65 @@ def c2_native_run_net(init_net, predict_net, inputs):
...
+        model(*inputs)
+        te = time.time()
+    total_pytorch_time = 0.0
+    for _i in range(main_iters):
+        ts = time.time()
+        model(*inputs)
+        te = time.time()
+        total_pytorch_time = te - ts + total_pytorch_time
```
Comment (dzhulgakov) [inline onnx_caffe2/helper.py]: += :)

File: onnx_caffe2/helper.py [hunk 10]
```diff
@@ -60,17 +68,65 @@ def c2_native_run_net(init_net, predict_net, inputs):
...
+    log.info("The PyTorch model execution time per iter is {} milliseconds, "
```
Comment (dzhulgakov) [inline onnx_caffe2/helper.py]: also return time?

File: examples/pytorch_to_caffe2.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+log.info("Save the Caffe2 models as pb files.")
+init_file = "./mymodel_init.pb"
+predict_file = "./mymodel_predict.pb"
+save_caffe2_net(init_net, init_file, output_txt=False)
+save_caffe2_net(predict_net, predict_file, output_txt=True)
+log.info("Load the Caffe2 models back.")
+init_net = load_caffe2_net(init_file)
+predict_net = load_caffe2_net(predict_file)
+
+# Prepare the inputs for Caffe2.
```
Comment (dzhulgakov) [inline examples/pytorch_to_caffe2.py]: this should be handled by passing option to Caffe2Backend

Comment (houseroad) [reply examples/pytorch_to_caffe2.py]: Yep.

File: onnx_caffe2/helper.py (hunk 7 above)
Comment (houseroad) [reply onnx_caffe2/helper.py]: init_net contains the fake input. So the assertion len(uninitialized) == len(inputs) fails. So it's better to pass a dict.

File: onnx_caffe2/helper.py (hunk 8 above)
Comment (houseroad) [reply onnx_caffe2/helper.py]: Nice catch.

File: onnx_caffe2/helper.py (hunk 9 above)
Comment (houseroad) [reply onnx_caffe2/helper.py]: Sure. :-)

File: onnx_caffe2/helper.py (hunk 10 above)
Comment (houseroad) [reply onnx_caffe2/helper.py]: Good point!

File: onnx_caffe2/helper.py (hunk 7 above)
Comment (dzhulgakov) [reply onnx_caffe2/helper.py]: What I meant is to accept the list directly in c2_native_run_net() and put this code there

File: onnx_caffe2/helper.py [hunk 12]
```diff
@@ -60,17 +68,73 @@ def c2_native_run_net(init_net, predict_net, inputs):
...
+    '''
+        Run the benchmark net on the target model.
+        Return the execution time per iteration (millisecond).
+    '''
+    ws = Workspace()
+    if init_net:
+        ws.RunNetOnce(init_net)
+    ws.CreateNet(predict_net)
+    results = ws.BenchmarkNet(predict_net.name, warmup_iters, main_iters, layer_details)
+    return results[0]
```
Comment (dzhulgakov) [inline onnx_caffe2/helper.py]: add `del ws` to free up memory explicitly
